### ISSUE TITLE
Add unique id to the web form element in order to support screen reader

### ISSUE THREAD

**Issue by fendiaoxiaoshuazi** [enketo/enketo-transformer#89]
**Description**
Our customers keep pushing us to make the Survey123 webform fully WCAG/508 compliant since this is a showstopper for them, e.g. the webform used on the government website needs to be fully WCAG/508 compliant.

We have made lots of effort to make the Survey123 keyboard navigatable, and we will need to keep going to make Survey123 support screen reader. To fully support the screen reader, most of the form elements need to have a unique id, and then we can use aria-labelledby, aira-describedby, and etc. to make the screen reader behave as expected. 

enketo already has aria-label added for elements, but it is not fully WCAG compliant, especially, we cannot fulfill the requirements in the VPAT, here is an example in the screenshot below:

![image](https://user-images.githubusercontent.com/26078106/102059051-58d80500-3e2b-11eb-99b5-e3c04e83577b.png)

Ref: Survey123 Web App VPAT
https://www.esri.com/content/dam/esrisites/en-us/media/legal/vpats/survey123-for-arcgis-311-vpat.pdf

**Expected behavior**
The unique id is needed for the label, description, input element, error msg, hint msg, etc.

Here is the demonstration for adding aria labels:

![7d7e9100-5ffe-11ea-82d1-d1cd545b19ad](https://user-images.githubusercontent.com/26078106/102059085-62fa0380-3e2b-11eb-9502-c5680fcbd52a.png)

**Comment by fendiaoxiaoshuazi** [comment]
@MartijnR Could you please take a look at this issue? The unique id is necessary for the screen reader support, and it becomes more and more important for the enterprise customer. 

Is it possible to make some change in the enketo to include the unique id for the form elements?

**Comment by MartijnR** [comment]
Thanks! I'd open to a PR for this, but we'd need to decide on some things first. I think it makes sense to do this in enketo-transformer and afterwards we'd have to go through all widgets in enketo-core (and survey123) to copy any relevant attributes (mostly on the `<input>`, `<select>`, `<textarea>` elements with perhaps some exceptions.

Let's figure exactly what attributes to add for all the enketo-core question elements:
* label
* required asterisk
* hint
* actual form control
* any sub-labels a widget might add (e.g. in the geopoint widget for latitude, longitude, search etc) though this would not need to be done in enketo-transformer
* error message (constraint and required)

Also, we need to figure out the approach for IDs on questions inside a repeat.

**Comment by fendiaoxiaoshuazi** [comment]
Hi @MartijnR, 

**Pattern of IDs:**
IDs are needed for all text elements, such as labels, hints, sub-labels, constraint msgs, and required msgs, since all text elements should be able to be referenced separately by the actual form control, and all text content should be announced by the screen reader.
For the ID patterns, we do need to consider repeat and probably multiple languages also. Do you have any idea or suggestion on that? My initial thought is that we generate a random and unique number like `679264` for each question, group, repeat (questions and groups in the repeat will get a different number or get `-posNumber-` appended), and then we append `label-`, `hint-`, `sublabel-`, `constraint-langCode-`, `required-langCode-` to form the id. @yixi8524 @nanzhang01 any thoughts on this?

**Here is the list of basic aria labels that are needed:**
- Add headings：
    - role=“heading” aria-level=“1” -> div.form-header (equals to <header>, must have one and only one)
    - role=“heading” aria-level=“2” -> legend.group-title>span.question-label
    - role=“heading” aria-level=“3” -> label>span.question-label
- Add following labels to all questions
    - Add aria-labelledby to input
    - Add aria-describedby to input
    - Add aria-required to required question
    - Add role=“alert” to error msg <span>
- Add role to applicable questions
    -  role="radiogroup"
- For the group expend/collapes icon
    - Add aria-expended aria-collpesed
    - Add description to demonstrate the function
- Handle the page information 
https://www.w3.org/WAI/tutorials/forms/multi-page/
    - Announce the page info when swiching the page
    - Add altnative text for the page indicator progress bar
    - Add description language to pervious and next page button e.g. "Move the next page of A11Y survey"
- For submit button, anounce the submitting status or unavailble stauts
- For group add/delete button, announce add and delete status

**Here is the expected condition for different question type:**
Single & Multiple & Likert:
![image](https://user-images.githubusercontent.com/26078106/102601411-1714ca80-415b-11eb-9a53-cc97b517a172.png)

Single choice grid:
![image](https://user-images.githubusercontent.com/26078106/102601431-1ed46f00-415b-11eb-8b3f-856e209463aa.png)

**Comment by MartijnR** [comment]
Thanks @fendiaoxiaoshuazi!

Some initial comments:

* There is also a second type of hint, called a guidance hint.
* Appending the type to the unique ID seems like a good way to do this.
* This does seem to significantly increase the size of the DOM, but I hope impact on performance will be acceptable.

Some of this will have to be taken care of by enketo-core (and some by the apps like enketo-express and survey123), but good to have this inventory here and then create different issues in those repos.

* wrt repeats, enketo-core already has built-in logic to keep track of ordinals that we could use to append to ids. These ordinals will remain constant (also if e.g. the 2nd repeat instance from a series of 4 is deleted).
* wrt widgets that hide the original form control (input, select, textarea) and replace it, could the attributes be copied to the widget? They won't be unique anymore, but the original form control would always stay hidden.
* itemsets (dynamic options) may be a bit of a pain, but at least won't have to be done in XSL...

**Comment by yixi8524** [comment]
@fendiaoxiaoshuazi @nanzhang01 and I will prepare an PR for this, test this new feature and the DOM performance. will need more discuss for this, thanks for you help.
